src/icastats_shared.c

```diff
@@ -46,74 +44,77 @@ volatile int stats_shm_handle = NOT_INITIALIZED;
 
 int stats_mmap(int user)
 {
+	int stats_shm_handle;
 	char shm_id[NAME_LENGHT];
 	struct stat stat_buf;
 
-	if (stats == NULL) {
-		sprintf(shm_id, "icastats_%d",
-			user == -1 ? geteuid() : (uid_t)user);
+	if (stats != NULL)
+		return 0;
 
-		stats_shm_handle = shm_open(shm_id, O_RDWR,  S_IRUSR | S_IWUSR);
-		if (stats_shm_handle == NOT_INITIALIZED)
-			stats_shm_handle = shm_open(shm_id, O_CREAT | O_RDWR,
-						    S_IRUSR | S_IWUSR);
+	sprintf(shm_id, "icastats_%d",
+		user == -1 ? geteuid() : (uid_t)user);
 
-		if (stats_shm_handle == NOT_INITIALIZED)
-			return -1;
+	stats_shm_handle = shm_open(shm_id,
+				    O_CREAT | O_RDWR,
+				    S_IRUSR | S_IWUSR);
 
-		if (user > 0 && geteuid() == 0) {
-			if (fchown(stats_shm_handle, user, user) == -1) {
-				close(stats_shm_handle);
-				return -1;
-			}
-		}
+	if (stats_shm_handle == -1)
+		return -1;
 
-		if (fstat(stats_shm_handle, &stat_buf)) {
-			close(stats_shm_handle);
-			return -1;
-		}
+	if ((user > 0 && geteuid() == 0) &&
+	    (fchown(stats_shm_handle, user, user) == -1))
+		goto err;
 
-		if (ftruncate(stats_shm_handle, STATS_SHM_SIZE) == -1) {
-			close(stats_shm_handle);
-			return -1;
-		}
+	if (fstat(stats_shm_handle, &stat_buf))
+		goto err;
 
-		stats = (stats_entry_t *) mmap(NULL, STATS_SHM_SIZE, PROT_READ |
-						 PROT_WRITE, MAP_SHARED,
-						 stats_shm_handle, 0);
-		if (stats == MAP_FAILED){
-			close(stats_shm_handle);
-			stats = NULL;
-			return -1;
-		}
+	if (ftruncate(stats_shm_handle, STATS_SHM_SIZE) == -1)
+		goto err;
+
+	stats = (stats_entry_t *) mmap(NULL, STATS_SHM_SIZE, PROT_READ |
+					 PROT_WRITE, MAP_SHARED,
+					 stats_shm_handle, 0);
 
-		if (stat_buf.st_size != STATS_SHM_SIZE)
-			memset(stats, 0, STATS_SHM_SIZE);
+	if (stats == MAP_FAILED){
+		stats = NULL;
+		goto err;
 	}
+
+	if (stat_buf.st_size != STATS_SHM_SIZE)
+		memset(stats, 0, STATS_SHM_SIZE);
+
+	close(stats_shm_handle);
 	return 0;
+
+err:
+	close(stats_shm_handle);
+	return -1;
 }
 
-/* Close and/or delete the shared memory segment
+/* unmap and (optionally) delete the shared memory segment for statistics
  * Argument:
+ * @user: uid for the shm segment handle. If user is -1, the effective uid is used.
  * @unlink - if unlink is true the shared memory segment will be
  * deleted. If it is false it will only be closed.
  */
 
-void stats_munmap(int unlink)
+void stats_munmap(int user, int unlink)
 {
-	char shm_id[NAME_LENGHT];
-	sprintf(shm_id, "icastats_%d", geteuid());
 
 	if (stats == NULL)
 		return;
 
 	munmap(stats, STATS_SHM_SIZE);
-	close(stats_shm_handle);
-	stats_shm_handle = NOT_INITIALIZED;
+	stats = NULL;
+
+	if(unlink == SHM_DESTROY) {
+		char shm_id[NAME_LENGHT];
+
+		sprintf(shm_id, "icastats_%d",
+			user == -1 ? geteuid() : (uid_t)user);
```

Comment:
Hopefully we are not setuid when calling this...  But in that case, simply do not call this function with -1 for `user` and `SHM_DESTROY` for unlink.

Reply:
The more I think about the syslog stuff, the more I come to the conclusion that we should remove all syslog() calls outside of the initialization. It is the responsibility of the caller of a library, to print/log/syslog errors. The library should only provide a possibility to detect a problem (e.g. by providing a error code). After removing all syslog() in rng_init() and drbg/set_error_state(), we can do a closelog() at the end of the initialization.

Reply:
I think that comment went to the wrong place.  Here we do nothing with syslog/closelog but re-create the name of the segment.

Reply:
uups

Reply:
Anyhow: I'll rework the syslog stuff and provide a v2.